### ISSUE TITLE
How to write a custom service and call methods from it?

### ISSUE THREAD

**Issue by TheOnlyArtz** [ballerina-platform/ballerina-lang#22588]
I'm writing a library in Ballerina and I'm looking for a way to emit events, just like the WebSocket's onError, onText etc'...

```bal
// library code
public type Client object {
    string token;
    http:WebSocketClient? wsConnection = ();
    service eventService;
    public function __init(string clientToken, service clientEventService) {
        self.token = clientToken;
        self.eventService = clientEventService;
        // I can't call any method which is inside the service because I'm not in a type descriptor (?)
    }
};
```

This is what I want to happen on the user's end:

```bal
Client myClient = new("TOKEN", myEventService);

service myEventService = service {
    resource function onTest(string myText) {
        io:println(myText);
    }
};

```

**Comment by daneshk** [comment]
@TheOnlyArtz Can we have more context about your use case. 

- What type of events need to be emitted to the user?
- How users use this object? 
- How this emit action is triggered. Is it from a client call?

**Comment by chamil321** [comment]
@TheOnlyArtz Ballerina Client/Listener/Service/Resource concepts are introduced along with a standard protocol. Eg. HTTP, WS, gRPC. Hence the calling the remote functions of service is happening according to the protocol spec. 

At the moment, Ballerina does not support service typing, hence the user cannot define custom services and write their own way of invoking its remote functions.

Also, I'm not clear of the `emit events` part of your question. Can you elaborate on it a little more?

**Comment by TheOnlyArtz** [comment]
@chamil321 Hey, and thanks for the response, what I mean by events is, in version 1.1.x of Ballerina there was this stream https://ballerina.io/v1-1/learn/by-example/streams.html which allowed well data streaming across the application, now it's impossible with the "new" stream? https://ballerina.io/v1-2/learn/by-example/streams.html

**Comment by mohanvive** [comment]
@TheOnlyArtz, sorry for the delayed response. BTW, I hope you are doing good. 

Let me give a detailed response to this 😄 
We are in the process of redesigning the streaming concepts of Ballerina. There will be two types of streams implementations in Ballerina, they are bounded streams (similar to java streams) and unbounded streams (which is used for event stream processing). 

In this redesigning process, we have removed the existing streaming implementation (which is available in Ballerina 1.1.x) in the latest Ballerina release. As of now, we have completed the bounded stream implementation and you could find those functionalities in the latest preview releases.   

AFAIU, you are looking for unbounded streaming capabilities and they will be available by mid next year (2021).

For now, if you are really looking for the event streaming capabilities then you could implement those using the existing Ballerina features such as lambda functions and etc but it involves some effort.

**Comment by mohanvive** [comment]
Answered, hence closing the issue.
